File: windows/src/GoalMaker.App/ViewModels/CalendarViewModel.cs
using System.Collections.ObjectModel;
using System.Globalization;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using GoalMaker.App.Localization;
using GoalMaker.Core.Planning;
using GoalMaker.Core.Settings;

namespace GoalMaker.App.ViewModels;

/// <summary>
/// The Calendar page (docs/calendar.md, spec story 68): a week or a month of planned tasks, deadlines
/// and reminders, with a day showing what it holds and a task opening from there.
/// </summary>
public sealed partial class CalendarViewModel : ObservableObject
{
    private readonly TaskList tasks;
    private readonly ReminderList reminders;
    private readonly ISettingsStore settings;
    private readonly IStrings strings;
    private readonly TimeProvider time;
    private readonly Action<string> openTask;
    private DateOnly? anchor;
    private DateOnly? selected;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsWeek))]
    [NotifyPropertyChangedFor(nameof(IsMonth))]
    private string kind = CalendarRules.Month;

    [ObservableProperty]
    private string period = string.Empty;

    [ObservableProperty]
    private string dayTitle = string.Empty;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasNoDay))]
    private bool hasDay;

    [ObservableProperty]
    private bool isDayEmpty;

    [ObservableProperty]
    private string dayReminders = string.Empty;

    [ObservableProperty]
    private bool hasReminders;

    public CalendarViewModel(
        TaskList tasks,
        ReminderList reminders,
        ISettingsStore settings,
        IStrings strings,
        TimeProvider time,
        Action<string> openTask,
        Action<Action> runOnUi)
    {
        this.tasks = tasks;
        this.reminders = reminders;
        this.settings = settings;
        this.strings = strings;
        this.time = time;
        this.openTask = openTask;
        tasks.Changed += (_, _) => runOnUi(Refresh);
        reminders.Changed += (_, _) => runOnUi(Refresh);
        Weekdays = [.. Enumerable.Range(0, 7).Select(day => CultureInfo.CurrentCulture.DateTimeFormat.AbbreviatedDayNames[(day + 1) % 7])];
        Refresh();
    }

    /// <summary>Which of the two views is on show, so its button reads as chosen.</summary>
    public bool IsWeek => Kind == CalendarRules.Week;

    public bool IsMonth => Kind == CalendarRules.Month;

    /// <summary>Whether no day is open, so the page asks the owner to pick one.</summary>
    public bool HasNoDay => !HasDay;

    /// <summary>The weekday headings, Monday first.</summary>
    public IReadOnlyList<string> Weekdays { get; }

    /// <summary>The days of the grid, seven to a row.</summary>
    public ObservableCollection<CalendarCellViewModel> Cells { get; } = [];

    /// <summary>What the day the owner picked holds.</summary>
    public ObservableCollection<CalendarEntryViewModel> DayEntries { get; } = [];

    public void Refresh()
    {
        var today = Today();
        var shown = anchor ?? today;
        var days = CalendarRules.Build(
            tasks.All(),
            reminders.All(),
            CalendarRules.Start(Kind, shown),
            CalendarRules.End(Kind, shown));

        Cells.Clear();
        foreach (var day in days)
        {
            var date = day.Day;
            Cells.Add(new CalendarCellViewModel(
                date,
                date.Day.ToString(CultureInfo.CurrentCulture),
                day.Count,
                date == today,
                Kind == CalendarRules.Week || date.Month == shown.Month,
                date == selected,
                () => Open(date)));
        }

        Period = Kind == CalendarRules.Month
            ? shown.ToString("MMMM yyyy", CultureInfo.CurrentCulture)
            : strings.Get(
                "Goals.Range",
                CalendarRules.Start(CalendarRules.Week, shown).ToString("d MMM", CultureInfo.CurrentCulture),
                CalendarRules.End(CalendarRules.Week, shown).ToString("d MMM", CultureInfo.CurrentCulture));

        var open = days.FirstOrDefault(day => day.Day == selected);
        DayEntries.Clear();
        if (open is not null)
        {
            foreach (var (task, label) in Entries(open))
            {
                var id = task.Id;
                DayEntries.Add(new CalendarEntryViewModel(
                    id,
                    task.Title,
                    strings.Get(label),
                    task.PlannedTime?.ToString("t", CultureInfo.CurrentCulture) ?? string.Empty,
                    task.State == TaskState.Done,
                    () => openTask(id)));
            }
        }

        DayTitle = open is null ? string.Empty : open.Day.ToString("D", CultureInfo.CurrentCulture);
        HasDay = open is not null;
        IsDayEmpty = open is not null && open.Empty;
        HasReminders = open is { Reminders: > 0 };
        DayReminders = HasReminders ? strings.Get("Calendar.Reminders", open!.Reminders) : string.Empty;
    }

    /// <summary>Shows the week or the month.</summary>
    [RelayCommand]
    public void Show(string? which)
    {
        Kind = which == CalendarRules.Week ? CalendarRules.Week : CalendarRules.Month;
        Refresh();
    }

    /// <summary>The week or month before the one on show.</summary>
    [RelayCommand]
    public void Back() => Step(-1);

    /// <summary>The week or month after the one on show.</summary>
    [RelayCommand]
    public void Forward() => Step(1);

    /// <summary>Back to the week or month holding today.</summary>
    [RelayCommand]
    public void Today_()
    {
        anchor = null;
        selected = null;
        Refresh();
    }

    /// <summary>Moves a task to another day, which is what dragging it onto a cell does.</summary>
    public void MoveTo(string taskId, DateOnly day)
    {
        tasks.Plan(taskId, day);
        selected = day;
        Refresh();
    }

    /// <summary>Opens a day, or closes it when it is already open.</summary>
    public void Open(DateOnly day)
    {
        selected = selected == day ? null : day;
        Refresh();
    }

    private static IEnumerable<(TaskItem Task, string Label)> Entries(CalendarDay day) =>
    [
        .. day.Planned.Select(task => (task, "Calendar.Planned")),
        .. day.Deadlines.Select(task => (task, "Calendar.Deadline")),
        .. day.Repeats.Select(task => (task, "Calendar.Repeat")),
    ];

    private void Step(int by)
    {
        var shown = anchor ?? Today();
        anchor = Kind == CalendarRules.Week ? shown.AddDays(7 * by) : shown.AddMonths(by);
        selected = null;
        Refresh();
    }

    private DateOnly Today() => PlanningDay.Of(time.GetLocalNow().DateTime, settings.DayStartHour);
}

File: android/app/src/main/kotlin/com/goalmaker/app/ui/calendar/CalendarViewModel.kt
package com.goalmaker.app.ui.calendar

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.goalmaker.app.application.planning.CalendarRules
import com.goalmaker.app.application.planning.ReminderList
import com.goalmaker.app.application.planning.TaskList
import com.goalmaker.app.application.settings.SettingsStore
import com.goalmaker.app.domain.planning.PlanningDay
import java.time.LocalDate
import java.time.LocalDateTime
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * The calendar (docs/calendar.md, spec story 68): a week or a month of planned tasks, deadlines and
 * reminders, with a day opening what it holds and a task moving to another day from there.
 */
class CalendarViewModel(
    private val tasks: TaskList,
    reminders: ReminderList,
    private val settings: SettingsStore,
    private val io: CoroutineDispatcher,
    private val clock: () -> LocalDateTime,
) : ViewModel() {
    private val view = MutableStateFlow(View(CalendarRules.MONTH, null, null))

    val uiState: StateFlow<CalendarUiState> = combine(
        tasks.watchAll().flowOn(io),
        reminders.watchAll().flowOn(io),
        view,
    ) { taskList, reminderList, showing ->
        val today = today()
        val anchor = showing.anchor ?: today
        CalendarUiState(
            loaded = true,
            kind = showing.kind,
            anchor = anchor,
            today = today,
            days = CalendarRules.build(
                taskList,
                reminderList,
                CalendarRules.start(showing.kind, anchor),
                CalendarRules.end(showing.kind, anchor),
            ),
            selected = showing.selected,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), CalendarUiState())

    /** Switches between the week and the month view, keeping the day in sight. */
    fun show(kind: String) = view.update { it.copy(kind = kind) }

    /** The week or month before the one on show. */
    fun back() = view.update { it.copy(anchor = step(it, -1), selected = null) }

    /** The week or month after the one on show. */
    fun forward() = view.update { it.copy(anchor = step(it, 1), selected = null) }

    /** Back to the week or month holding today. */
    fun today(reset: Boolean) = view.update { it.copy(anchor = null, selected = null) }

    /** Opens a day, or closes it when it is already open. */
    fun open(day: LocalDate) = view.update { it.copy(selected = if (it.selected == day) null else day) }

    /** Moves a task to another day, the way Plan tomorrow does. */
    fun plan(taskId: String, day: LocalDate) {
        viewModelScope.launch(io) { tasks.plan(taskId, day) }
    }

    private fun step(showing: View, by: Long): LocalDate {
        val anchor = showing.anchor ?: today()
        return if (showing.kind == CalendarRules.WEEK) anchor.plusWeeks(by) else anchor.plusMonths(by)
    }

    private fun today(): LocalDate = PlanningDay.of(clock(), settings.dayStartHour.value)

    private fun MutableStateFlow<View>.update(edit: (View) -> View) {
        value = edit(value)
    }

    private data class View(val kind: String, val anchor: LocalDate?, val selected: LocalDate?)
}
